### ISSUE TITLE
Why did you choose the softmax cross entropy ?

### ISSUE THREAD

**Issue by nianzu-ethan-zheng** [musyoku/adversarial-autoencoder#10]
Hi ,Thanks for your code ,  I am confused about why you choose the softmax cross entropy ,  I also tried this loss ,unfortunately , it did work but not so good , when I regularized the z representation space into a gaussian distribution using the softmax cross entropy ,  I got a off-centre distribution ,like the following:
![result](https://github.com/DreamPurchaseZnz/Picture/blob/master/AAE/unsupervised/regularized_z/gaussian/inherit%20from%20unifrom%20var%201/labeled_z.png)

you can see gaussian expectation is about (-2, -2.5) rather than (0 , 0) ,which quitely confused me. 
I can't explain it from theory, and you can see the equilibrium between the generator and discriminator,

![equilibrium](https://github.com/DreamPurchaseZnz/Picture/blob/master/AAE/unsupervised/regularized_z/gaussian/inherit%20from%20unifrom%20var%201/Loss-Adversarial-z.png)

D and G have reached the expected value about log2 and log4 respectively, and  I don't know why the equilibrium  can be found,  In general standpoint,  discriminator can be easily  judge the sample generated from generator, and I have been well check the code , it have no bug or type the wrong variable. I don't know what's going on and what't wrong with it ?

**Comment by musyoku** [comment]
Have you plotted *representation* rather than *z* ?

![screenshot from 2017-10-09 22-16-29](https://user-images.githubusercontent.com/15250418/31340044-e2198314-ad3f-11e7-90e4-4ca07e614a05.png)

Since representation is the sum of cluster-head and z, the center may not come to (0, 0).

### z

![scatter_z](https://user-images.githubusercontent.com/15250418/31340157-3ed2effa-ad40-11e7-9d40-9fc36d201a7e.png)
 
### representation

![scatter_r](https://user-images.githubusercontent.com/15250418/31340176-4816799c-ad40-11e7-8e5f-376ae298d54a.png)

**Comment by musyoku** [comment]
In the latest version of code, the definition of discriminator for z is as follows:

`run/unsupervised/dim_reduction/model.py`

```
self.discriminator_z = nn.Module(
	nn.GaussianNoise(std=0.3),   # add noise
	nn.Linear(ndim_z, ndim_h),
	nn.ReLU(),
	nn.Linear(ndim_h, ndim_h),
	nn.ReLU(),
	nn.Linear(ndim_h, 2),
)
```
If you delete `nn.GaussianNoise(std=0.3),`, discriminator can easily judge the samples generated from generator.

**Comment by nianzu-ethan-zheng** [comment]
Yeah Thanks your reply.  I just regularize the representation z without the transform , The encoder just output 2 dimensional vector which just are put into discriminator. The lastest result actually show that Maybe the training epoch is too much, Let me expian it:

z space just like the below in  the earlier epoch 
![earlier](https://github.com/DreamPurchaseZnz/Picture/blob/master/AAE/unsupervised/regularized_z/uniform_desk_radius1_sigmoid/labeled_z.png)

when you try to train more epoch in oder to  get more clearly boundary, the result just like this:
![later](https://github.com/DreamPurchaseZnz/Picture/blob/master/AAE/unsupervised/regularized_z/uniform_desk_radius1_sigmoid/more_epoch/labeled_z.png)

you can see the the whole space just move left with slightly changing the shape. 
The gaussian noise of  input layer of discriminator has been removed .

So I guess the adversarial loss maybe give in to the autoencoder loss, while the adversarial loss just change slightly(about 0.001)  and the autoencoder loss decreasing obviously.(about 0.1 - 1) 

So the autoencoder pull the space to off-centre , is right ?

**Comment by musyoku** [comment]
I have the same view as you.

**Comment by musyoku** [comment]
What happens if you try z to 10 2d-gaussian mixture or swiss-roll ?

![scatter_gen](https://user-images.githubusercontent.com/15250418/31341580-e7a011e0-ad44-11e7-970f-77e8939e1661.png)

![swiss_roll](https://user-images.githubusercontent.com/15250418/31341603-f928a0bc-ad44-11e7-8781-a55100982805.png)

**Comment by nianzu-ethan-zheng** [comment]
I think the result you posted is under the condition of semi supervised learning, I alse tried the unsupervised learning, giving the result as follows:
![ Swiss row -label 10](https://github.com/DreamPurchaseZnz/Picture/blob/master/AAE/unsupervised/regularized_z/swiss_roll_label10/labeled_z.png)

the gaussian mixture like this 
![Gaussian mixture -label 10](https://github.com/DreamPurchaseZnz/Picture/blob/master/AAE/unsupervised/regularized_z/gaussian_mixture_10/epoch_175/labeled_z.png)

**Comment by nianzu-ethan-zheng** [comment]
I think It's hard to constrain the representation z of  unsupervise learning  to a certain area, such as (-2,2). the space drift about to get better interpretation of real space. But the most confused thing is the space  magically keep the shape and structure.

**Comment by musyoku** [comment]
I tried the unsupervised learning.
The results after 300 epochs is as follows:

10 2d-gaussian mixture
![gauss](https://user-images.githubusercontent.com/15250418/31485333-7134c420-af6e-11e7-8438-709c2a83ca90.png)

10 2d-gaussian mixture + noise
![gauss_noise](https://user-images.githubusercontent.com/15250418/31485371-897adf38-af6e-11e7-9405-590de98f6a38.png)

swissroll
![swiss](https://user-images.githubusercontent.com/15250418/31485381-937fa914-af6e-11e7-87fe-97b887781c2a.png)

swissroll + noise
![swiss_noise](https://user-images.githubusercontent.com/15250418/31485390-9bb58e6e-af6e-11e7-8b63-32eaf7f7d004.png)

code:
https://gist.github.com/musyoku/ad203d1cb24c60b0926043125e676c71

**Comment by nianzu-ethan-zheng** [comment]
```
[y_onehot_u = xp.zeros((1, model.ndim_y), dtype=xp.float32)
y_onehot_u[0, -1] = 1	# turn on the extra class
y_onehot_u = xp.repeat(y_onehot_u, args.batchsize, axis=0)]
```

I think the y_onehot_u is unnecessary. Maybe I find the origin of causing the space shifting , which  batch normalization can account for . In the later experiment when I remove BN , the representation space bound to the ideal area. But there is another problem followed that the autoencoder the adversarial 
phase suddenly break the balance between encoder and discriminator ,you can see the plot as followed:

![loss-adversarial-z](https://user-images.githubusercontent.com/26301441/31549429-bf2fb066-b060-11e7-8f17-cf356278ecd0.png)

and the reconstruction loss react to it.

![loss-reconstruction](https://user-images.githubusercontent.com/26301441/31549497-f5b9af56-b060-11e7-8e39-2db4c3974799.png)

Do you early stopping your training ?

**Comment by musyoku** [comment]
No I don't.
I think that the reconstruction loss value in your plot is too high.
In my experiment the reconstruction loss is less than 0.1.

I found that BN didn't work, so I don't use BN in all my experiments.
I also found that adding gaussian noise to the input of the discriminator stabilizes GAN.

**Comment by nianzu-ethan-zheng** [comment]
I don't know why the reconstruction is too high, Maybe it's a programming problem, the formula is as follow:
![image](https://user-images.githubusercontent.com/26301441/31601890-ec74618c-b28d-11e7-8eb9-1074ad3674d3.png)

and using the same structure and same parameters ,such as learning rate =0.0001, momentum=0.5, however the balance will be broken while I train the network more than 300 epochs . I don't know why and I have been tried adding some noise to the input of discriminator (std =0.3/0.01,I think it's too big, so I tried std = 0.01), but it don't work, By the way , I implement the idea through tensorflow instead of chainer.

**Comment by musyoku** [comment]
Please try the supervised learning with only reconstruction loss.
set the learning rate to a high value and check whether the training reconstruction loss becomes a small value (or 0).
If not, the implementation is incorrect.
